### PR TITLE
Add Argument to SchemaModel

### PR DESCRIPTION
**Pull request by a-alle**
# Description

This PR introduces the concept of Argument in the SchemaModel.
It is currently present on Attribute as an property of type Argument[]

## Complexity

Complexity: Low

# Issue

> **Note**
>
>  Please link to the GitHub issue(s) in which the proposal for this work was discussed
>  
>  To link to multiple issues, use full syntax for each, for example `Closes #1, closes #2, closes #3`

Closes 

# Checklist

The following requirements should have been met (depending on the changes in the branch):

- [ ] Documentation has been updated
- [ ] TCK tests have been updated
- [ ] Integration tests have been updated
- [ ] Example applications have been updated
- [ ] New files have copyright header
- [ ] CLA (https://neo4j.com/developer/cla/) has been signed

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
Made some changes to the Neo4j GraphQL Toolbox? Or made some changes to the core library perhaps?

Check out your changes at https://neo4j-graphql-toolbox-3883.surge.sh

Use for instance a local dbms hosted on [Neo4j Desktop](https://neo4j.com/download/) or a cloud-hosted [Aura Free](https://neo4j.com/cloud/aura-free/) dbms.

**Comment by neo4j-team-graphql** [comment]
# Performance Report

No Performance Changes


<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3403 | 3403 | 37 | 67 | 1134 |
|  aggregations.NestedAggregation | 16553 | 16553 | 87 | 114 | 2174 |
|  aggregations.AggregationWithWhere | 11979 | 11979 | 53 | 68 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 22116987 | 22116987 | 3458 | 3501 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 22116987 | 22116987 | 3109 | 3255 | 2008534 |
|  aggregations.NestedCountFromMovieToActors | 9734 | 9734 | 42 | 59 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 9937 | 9937 | 54 | 69 | 2174 |
|  aggregations.DeeplyNestedCount | 13070331 | 13070331 | 5243 | 5253 | 2008534 |
|  batch-create.BatchCreate | 4200 | 4200 | 193 | 202 | 600 |
|  batch-create.BatchCreateSmall | 77 | 77 | 79 | 106 | 11 |
|  connect.createAndConnect | 14424 | 14424 | 166 | 190 | 3003 |
|  connections.Connection | 14082 | 14082 | 62 | 81 | 2174 |
|  connections.NestedConnection | 45470 | 45470 | 127 | 143 | 4516 |
|  connections.ConnectionWithSort | 3284 | 3284 | 73 | 80 | 1040 |
|  connections.ConnectionWithSortAndCypher | 16325 | 16325 | 104 | 132 | 2174 |
|  create.SimpleMutation | 7 | 7 | 64 | 77 | 1 |
|  cypher-directive.TopLevelMutationDirective | 1135 | 1135 | 31 | 31 | 1134 |
|  delete.SimpleDelete | 19401 | 19401 | 279 | 239 | 1040 |
|  delete.NestedDeleteInUpdate | 19198 | 19198 | 167 | 215 | 1179 |
|  2925.SingleRelationshipFilter | 6468 | 6468 | 63 | 60 | 1040 |
|  2925.NestedSingleRelationshipFilter | 22900 | 22900 | 112 | 138 | 2174 |
|  2925.SingleRelationshipRequiredFilter | 5201 | 5201 | 50 | 56 | 1040 |
|  2925.NestedSingleRelationshipRequiredFilter | 9361 | 9361 | 70 | 73 | 1040 |
|  query.SimpleQuery | 3121 | 3121 | 35 | 36 | 1040 |
|  query.SimpleQueryWithRelationship | 16162 | 16162 | 53 | 74 | 2174 |
|  query.QueryWhere | 9703 | 9703 | 42 | 56 | 2163 |
|  query.SimpleQueryWithNestedWhere | 9881 | 9881 | 59 | 89 | 2163 |
|  query.Nested | 10096041 | 10096041 | 11299 | 11066 | 2008534 |
|  query.NestedWithFilter | 10074401 | 10074401 | 11154 | 11053 | 2004000 |
|  query.OrFilterOnRelationships | 43469 | 42934 | 230 | 239 | 2128 |
|  query.OrFilterOnRelationshipsAndNested | 37383 | 37519 | 364 | 336 | 2128 |
|  query.QueryWithNestedIn | 13729 | 13716 | 82 | 74 | 1326 |
|  query.NestedConnectionWhere | 9834 | 9834 | 71 | 77 | 2174 |
|  query.DeeplyNestedConnectionWhere | 10054 | 9974 | 96 | 135 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 19350 | 19260 | 146 | 178 | 1664 |
|  query.NestedWithRelationshipSingleFilters | 3881 | 3881 | 198 | 246 | 1134 |
|  query.Fulltext | 80 | 80 | 36 | 42 | 16 |
|  query.FulltextWithNestedQuery | 587 | 587 | 48 | 63 | 84 |
|  sorting.SortMultipleTypes | 3515 | 3515 | 74 | 140 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 14501 | 14522 | 108 | 131 | 2174 |
|  sorting.SortOnNestedFields | 14082 | 14082 | 74 | 118 | 2174 |
|  sorting.SortDeeplyNestedFields | 43198 | 43198 | 98 | 174 | 4516 |
|  sorting.SortWithTopLevelCypher | 3396 | 3396 | 69 | 81 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 69 | 75 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 75 | 72 | 35 |
|  unions.NestedUnion | 410637 | 410637 | 412 | 372 | 33033 |
|  unions.NestedUnionWithMissingFields | 384611 | 384611 | 325 | 298 | 33033 |
|  update.NestedUpdate | 16143 | 16143 | 129 | 124 | 2002 |

</details>
        
Old Schema Generation: 49.668s
Schema Generation: 51.112s
Old Subgraph Schema Generation: 57.130s
Subgraph Schema Generation: 59.995s

File: packages/graphql/src/schema-model/argument/Argument.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type { ValueNode } from "graphql";
+import { parseValueNode } from "../parser/parse-value-node";
+import type { AttributeType } from "../attribute/AttributeType";
+
+export class Argument {
+    // these are arguments (eg field arguments)
```
Comment (angrykoala) [inline packages/graphql/src/schema-model/argument/Argument.ts:25]: I'm not sure this comment gives useful context

File: packages/graphql/src/schema-model/argument/Argument.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    public readonly name: string;
+    public readonly type: AttributeType;
+    public readonly defaultValue?: string;
+    public readonly description: string;
```
Comment (angrykoala) [inline packages/graphql/src/schema-model/argument/Argument.ts:28]: I don't think these properties are used in the translation layer. Are these needed in the augmented schema?

File: packages/graphql/src/schema-model/argument/Argument.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    // public readonly annotations: Partial<Annotations> = {};
```
Comment (angrykoala) [inline packages/graphql/src/schema-model/argument/Argument.ts:30]:
```suggestion
```

File: packages/graphql/src/schema-model/argument/model-adapters/ArgumentAdapter.ts [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+        this.description = argument.description;
+        this.assertionOptions = {
+            includeLists: true,
+        };
+    }
+
+    /**
+     * Just a helper to get the wrapped type in case of a list, useful for the assertions
+     */
+    private getTypeForAssertion(includeLists: boolean) {
```
Comment (angrykoala) [inline packages/graphql/src/schema-model/argument/model-adapters/ArgumentAdapter.ts:59]: What does it mean `ForAssertion` in this method?

File: packages/graphql/src/schema-model/argument/model-adapters/ArgumentAdapter.ts [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+        if (includeLists) {
+            return this.isList() ? this.type.ofType : this.type;
+        }
+        return this.type;
+    }
+
+    isBoolean(options = this.assertionOptions): boolean {
```
Comment (angrykoala) [inline packages/graphql/src/schema-model/argument/model-adapters/ArgumentAdapter.ts:66]: I feel I've seen this before in other adapters

Are we sure this is not duplicate code?

Comment (MacondoExpress) [inline packages/graphql/src/schema-model/argument/model-adapters/ArgumentAdapter.ts:66]: Not blocking:  This duplicated code, maybe can be solved with https://www.typescriptlang.org/docs/handbook/mixins.html with a type-helper class, or maybe we could move outside these helpers directly outside. But I would want to revisit the "adapter" idea later in maybe just using directly the Schema Model and having an external GraphQL view to deal with the presentation logic for the typesName and helpers function outside, but we can properly design this later!

**Review by MacondoExpress — DISMISSED** [review]

File: packages/graphql/src/schema-model/argument/Argument.ts (hunk 2 above)
Comment (a-alle) [reply packages/graphql/src/schema-model/argument/Argument.ts:28]: yes

File: packages/graphql/src/schema-model/argument/model-adapters/ArgumentAdapter.ts (hunk 5 above)
Comment (a-alle) [reply packages/graphql/src/schema-model/argument/model-adapters/ArgumentAdapter.ts:66]: so is anything in the scope of this PR?

Comment (MacondoExpress) [reply packages/graphql/src/schema-model/argument/model-adapters/ArgumentAdapter.ts:66]: No, in my opinion is not worth spending time now to clean up this code if we already agreed that is not the final solution.

**Pull request merged**